man/man1/lpj.1
.TH lpjml 1  "October 28, 2008" "version 3.4.017" "USER COMMANDS"
.SH NAME
lpj, lpjml \- Dynamic global vegetation model with managed land use and river routing
.SH SYNOPSIS
.B lpjml
[\-h] [\-v] [-vv] [-fpe] [\-output {gather|mpi2|buffer}] [\-outpath \fIdir\fP]
[\-inpath \fIdir\fP] [\-restartpath \fIdir\fP]
[[\-Dmacro[=value]] [\-I\fIdir\fP] ...] [\fIfilename\fP]

.B lpj
[\-h] [\-v] [-vv] [-fpe] [\-output {gather|mpi2|buffer}] [\-outpath \fIdir\fP]
[\-inpath \fIdir\fP] [\-restartpath \fIdir\fP]
[[\-Dmacro[=value]] [\-I\fIdir\fP] ...] [\fIfilename\fP]
.SH DESCRIPTION
The model simulates vegetation dynamics, hydrology and soil 
organic matter dynamics on an area-averaged grid cell basis using
one-year time step. The name derives from the three locations \fBL\fPund-\fBP\fPotsdam-\fBJ\fPena but is no longer to be interpreted that way. Input parameters are monthly mean air
temperature, total precipitation and percentage of full sunshine,
annual atmospheric CO2 concentration and soil texture class. The
simulation for each grid cell begins from "bare ground",
requiring a "spin up" (under non-transient climate) of c. 1000
years to develop equilibrium vegetation and soil structure. A restart
file can be created to save compute time for the spin-up phase. 
.SH OPTIONS
.TP
\-h
display a short help text
.TP
\-v
print information about program version, operating system, compiler used, and compiler flags set.
.TP
\-vv
verbosely print the actual values during reading of the configuration files.
.TP
\-fpe
enable floating point exceptions. A core file is generated in case of an arithmetic error. Option is only available if \fBlpjml\fP has been compiled with -DWITH_FPE flag.
.TP
\-output {gather|mpi2|buffer}
use the given output method to write data to disk. Valid output methods are
gather, mpi2, and buffer. The default output method is gather sending all output to the root task. If memory allocation fails use other output method. Method mpi2 uses MPI-2 parallel file functions, while method buffer stores all output in memory.  If memory allocation fails for buffer use other output method. The option is only valid for the MPI version of the code.  
.TP
\-outpath \fIdir\fP
set the output directory path. The path is added to the output filenames if they do not contain an absolute path.
.TP
\-inpath \fIdir\fP
set the input directory path. The path is added to the input filenames if they do not contain an absolute path.
.TP
\-restartpath \fIdir\fP
set the restart directory path. The path is added to the restart filenames if they do not contain an absolute path.
.TP
\-Dmacro[=value]
define macro for the preprocessor of the configuration file
.TP
\-I\fIdir\fP
define include directory for the preprocessor of the configuration file
.TP
.I filename
name of configuration file, default is \fIlpjml.conf\fP for \fBlpjml\fP and \fIlpj.conf\fP for \fBlpj\fP.
.SH EXAMPLES
.TP
Perform simulation with input from the /scratch file system:
.B lpjml
\-inpath /scratch/02/lpj
.PP
.TP
Start simulation from restart file with floating point exceptions enabled:
.B lpjml
\-DFROM_RESTART \-fpe lpjml.conf
.PP
For running the MPI version of the program on more than one task the
.B lpjsubmit
script has to be used.
.SH ENVIRONMENT
.TP
LPJCONFIG
default LPJmL configuration filename
.TP
LPJPREP 
defines preprocessor command for LPJmL configuration file, default is \fBcpp -P\fP.
.TP
LPJROOT
defines the root directory for LPJmL. This directory is added to the
include directories of the preprocessor. Invoking \fBlpj_paths.sh\fP or \fBlpj_paths.csh\fP will set this
environment variable correctly.
.TP
LPJINPATH
path appended to the input filenames. Only done for filenames without absolute path. Same as '-inpath \fIdir\fP' option.
.TP
LPJOPTIONS     
runtime options for the preprocessor of LPJmL configuration files
.TP
LPJOUTPATH
path appended to the output filenames. Only done for filenames without absolute path. Same as '-outpath \fIdir\fP' option.
.TP
LPJRESTARTPATH
path appended to the restart filenames. Only done for filenames without absolute path. Same as '-restartpath \fIdir\fP' option.
.TP
LPJOUTPUT
default method for generating output files. Valid values
are mpi2, buffer, gather. Used by the MPI version only. Same as '-output method' option.

.SH EXIT STATUS
.B lpjml
and
.B lpj 
returns a zero exit status if the simulation finishes successfully.
Non zero is returned in case of failure.
.SH DIAGNOSTICS
On error the program prints the following message on stderr:

ERRORxxx: message, program terminated unsuccessfully

where xxx is the return code of the program. The following error codes are defined:
.TP
ERROR001 
Error reading configuration. Check syntax of configuration file. 
.TP
ERROR002 
Error initializing climate. Check whether all input files are present.
.TP
ERROR003 
Error initializing land use, check for missing input file.
.TP
ERROR004 
Error initializing water use, check for missing input file.
.TP
ERROR005 
Error initializing grid, check for missing input files.
.TP
ERROR006 
Error initializing drainage, check for missing input files.
.TP
ERROR007 
Invalid carbon balance. Internal error
.TP
ERROR008 
Invalid water balance. Internal error
.TP
ERROR009 
Negative discharge. Internal error
.TP
ERROR010 
Negative fire probability. Internal error.
.TP
ERROR011 
Negative soil moisture. Internal error
.TP
ERROR012 
Error allocating memory. Rerun parallel program on more MPI tasks to reduce memory per task.
.TP
ERROR013 
Negative stand fraction. Internal error.
.TP
ERROR014 
Stand fraction sum error. Can be caused by invalid restart file.
.TP
ERROR015 
List is empty in dellistitem. Internal error.
.TP
ERROR016 
Index out of range in dellistitem. Internal error
.TP
ERROR017
Error in newlanduse. Can be caused by invalid restart file.
.TP
ERROR018 
Invalid year in 'getco2'. CO2 data file is too short.
.TP
ERROR019 
Crop fraction >1. Internal error.
.TP
ERROR020 
No natural stand for deforest. Internal error.
.TP
ERROR021 
Wrong cultivation type. Internal error.
.TP
ERROR022 
Invalid discharge. Internal error.
.TP
ERROR023 
Cannot allocate buffer memory. Rerun program with option '-output mpi2' or '-output gather'. The total amount of memory allocated for the output buffers can be obtained by invoking \fBlpjcheck\fP. The buffer size has to be divided by the number of MPI tasks.
.TP
ERROR024 
PFT list is not empty in setaside. Internal error.
.TP
ERROR025 
Negative establishment rate. Internal error.
.TP
ERROR026 
Floating point error occurred. Floating point exceptions will only be thrown if -fpe option is set. This is in particular useful if NaNs appear in the output files.  Internal error.
.PP
Internal errors will generate a core dump and have to be fixed by changes in the code. A "post-mortem" analysis can be made by calling 

\fBgdb\fP $LPJROOT/bin/lpjml core 

It is recommended to compile the code without optimization and inlining making the inspection of the core file easier. Configure in $LPJROOT with

.nf
\fBconfigure.sh\fP -debug
\fBmake\fP clean
.B make
.fi

will do the job. If no core file is generated set the user limit for core files:

\fBulimit\fP -c unlimited

Some of these errors are only raised if the -DSAFE flag has been set in \fIMakefile.inc\fP. The flags set at compile time can be obtained by invoking \fBlpjml -v\fP.

.SH AUTHORS

See AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
lpjsubmit(1), lpjcheck(1), cru2clm(1), txt2clm(1), grid2clm(1), cft2clm(1), lpjdistribute(1), lpjdistribute_river(1), lpjcat(1), lpjprint(1), printheader(1), cat2bsq(1), output_bsq(1), lpj_paths.sh(1), lpj_paths.csh(1), newpft.sh(1), lpjml.conf(5), lpj.conf(5), clm(5)
.TP
The LPJ & LPJmL Web Distribution Portal:
http://www.pik-potsdam.de/lpj



lpj.js
/**************************************************************************************/
/**                                                                                \n**/
/**                   l  p  j   .  j  s                                            \n**/
/**                                                                                \n**/
/** Configuration file for LPJmL C Version 5.1.001 without land use                \n**/
/**                                                                                \n**/
/** Configuration file is divided into five sections:                              \n**/
/**                                                                                \n**/
/**  I.   Simulation description and type section                                  \n**/
/**  II.  Input parameter section                                                  \n**/
/**  III. Input data section                                                       \n**/
/**  IV.  Output data section                                                      \n**/
/**  V.   Run settings section                                                     \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/
{
#include "include/conf.h" /* include constant definitions */

/* #define DAILY_OUTPUT */

/*===================================================================*/
/*  I. Simulation description and type section                       */
/*===================================================================*/

"sim_name" : "LPJ run with PNV",   /* Simulation description */
"sim_id"   : LPJ,         /* LPJML Simulation with natural vegetation only */
"version"  : "5.1",       /* LPJmL version expected */
"random_prec" : true,     /* Random weather generator for precipitation enabled */
"random_seed" : 2,        /* seed for random number generator */
"radiation" : RADIATION,  /* other options: CLOUDINESS, RADIATION, RADIATION_SWONLY, RADIATION_LWDOWN */
"fire" : FIRE,            /* fire disturbance enabled, other options: NO_FIRE, FIRE, SPITFIRE, SPITFIRE_TMAX */
#ifdef FROM_RESTART
"population" : false,      /* used by SPITFIRE model */
#else
"population" : false,
#endif
"prescribe_burntarea" : false,
"prescribe_landcover" : NO_LANDCOVER, 
"new_phenology": true,
"river_routing" : true,
"equilsoil" :false,
"permafrost" : true,
"with_nitrogen" : LIM_NITROGEN, /* other options: NO_NITROGEN, LIM_NITROGEN, UNLIM_NITROGEN */
"const_climate" : false,
"const_deposition" : false,

/*===================================================================*/
/*  II. Input parameter section                                      */
/*===================================================================*/

#include "param.js"    /* Input parameter file */

/*===================================================================*/
/*  III. Input data section                                          */
/*===================================================================*/

#include "input_crumonthly.js"    /* Input files of CRU dataset */

/*===================================================================*/
/*  IV. Output data section                                          */
/*===================================================================*/

#ifdef WITH_GRIDBASED
"pft_output_scaled" : GRIDBASED,
#else
"pft_output_scaled" : PFTBASED,
#endif

#ifdef FROM_RESTART

"output" : [

/*
ID                         Fmt                    filename
-------------------------- ---------------------- ----------------------------- */
{ "id" : GRID,             "file" : { "fmt" : RAW, "name" : "output/grid.bin" }},
{ "id" : FPC,              "file" : { "fmt" : RAW, "name" : "output/fpc.bin"}},
{ "id" : MNPP,             "file" : { "fmt" : RAW, "name" : "output/mnpp.bin"}},
{ "id" : MGPP,             "file" : { "fmt" : RAW, "name" : "output/mgpp.bin"}},
{ "id" : MRH,              "file" : { "fmt" : RAW, "name" : "output/mrh.bin"}},
{ "id" : MFAPAR,           "file" : { "fmt" : RAW, "name" : "output/mfapar.bin"}},
{ "id" : MTRANSP,          "file" : { "fmt" : RAW, "name" : "output/mtransp.bin"}},
{ "id" : MRUNOFF,          "file" : { "fmt" : RAW, "name" : "output/mrunoff.bin"}},
{ "id" : MEVAP,            "file" : { "fmt" : RAW, "name" : "output/mevap.bin"}},
{ "id" : MINTERC,          "file" : { "fmt" : RAW, "name" : "output/minterc.bin"}},
{ "id" : MSWC1,            "file" : { "fmt" : RAW, "name" : "output/mswc1.bin"}},
{ "id" : MSWC2,            "file" : { "fmt" : RAW, "name" : "output/mswc2.bin"}},
{ "id" : FIREC,            "file" : { "fmt" : RAW, "name" : "output/firec.bin"}},
{ "id" : FIREF,            "file" : { "fmt" : RAW, "name" : "output/firef.bin"}},
{ "id" : VEGC,             "file" : { "fmt" : RAW, "name" : "output/vegc.bin"}},
{ "id" : SOILC,            "file" : { "fmt" : RAW, "name" : "output/soilc.bin"}},
{ "id" : LITC,             "file" : { "fmt" : RAW, "name" : "output/litc.bin"}},
{ "id" : FLUX_ESTABC,      "file" : { "fmt" : RAW, "name" : "output/flux_estab.bin"}},
{ "id" : PFT_VEGC,         "file" : { "fmt" : RAW, "name" : "output/pft_vegc.bin"}},
{ "id" : MPHEN_TMIN,       "file" : { "fmt" : RAW, "name" : "output/mphen_tmin.bin"}},
{ "id" : MPHEN_TMAX,       "file" : { "fmt" : RAW, "name" : "output/mphen_tmax.bin"}},
{ "id" : MPHEN_LIGHT,      "file" : { "fmt" : RAW, "name" : "output/mphen_light.bin"}},
{ "id" : MPHEN_WATER,      "file" : { "fmt" : RAW, "name" : "output/mphen_water.bin"}},
{ "id" : VEGN,             "file" : { "fmt" : RAW, "name" : "output/vegn.bin"}},
{ "id" : SOILN,            "file" : { "fmt" : RAW, "name" : "output/soiln.bin"}},
{ "id" : LITN,             "file" : { "fmt" : RAW, "name" : "output/litn.bin"}},
{ "id" : SOILN_LAYER,      "file" : { "fmt" : RAW, "name" : "output/soiln_layer.bin"}},
{ "id" : SOILNO3_LAYER,    "file" : { "fmt" : RAW, "name" : "output/soilno3_layer.bin"}},
{ "id" : SOILNH4_LAYER,    "file" : { "fmt" : RAW, "name" : "output/soilnh4_layer.bin"}},
{ "id" : SOILN_SLOW,       "file" : { "fmt" : RAW, "name" : "output/soiln_slow.bin"}},
{ "id" : SOILNH4,          "file" : { "fmt" : RAW, "name" : "output/soilnh4.bin"}},
{ "id" : SOILNO3,          "file" : { "fmt" : RAW, "name" : "output/soilno3.bin"}},
{ "id" : PFT_NUPTAKE,      "file" : { "fmt" : RAW, "name" : "output/pft_nuptake.bin"}},
{ "id" : MNUPTAKE,         "file" : { "fmt" : RAW, "name" : "output/mnuptake.bin"}},
{ "id" : MLEACHING,        "file" : { "fmt" : RAW, "name" : "output/mleaching.bin"}},
{ "id" : MN2O_DENIT,       "file" : { "fmt" : RAW, "name" : "output/mn2o_denit.bin"}},
{ "id" : MN2O_NIT,         "file" : { "fmt" : RAW, "name" : "output/mn2o_nit.bin"}},
{ "id" : MN2_EMIS,         "file" : { "fmt" : RAW, "name" : "output/mn2_emis.bin"}},
{ "id" : MBNF,             "file" : { "fmt" : RAW, "name" : "output/mbnf.bin"}},
{ "id" : MN_IMMO,          "file" : { "fmt" : RAW, "name" : "output/mn_immo.bin"}},
{ "id" : PFT_NDEMAND,      "file" : { "fmt" : RAW, "name" : "output/pft_ndemand.bin"}},
{ "id" : FIREN,            "file" : { "fmt" : RAW, "name" : "output/firen.bin"}},
{ "id" : MN_MINERALIZATION,"file" : { "fmt" : RAW, "name" : "output/mn_mineralization.bin"}},
{ "id" : MN_VOLATILIZATION,"file" : { "fmt" : RAW, "name" : "output/mn_volatilization.bin"}},
{ "id" : PFT_NLIMIT,       "file" : { "fmt" : RAW, "name" : "output/pft_nlimit.bin"}},
{ "id" : PFT_VEGN,         "file" : { "fmt" : RAW, "name" : "output/pft_vegn.bin"}},
{ "id" : PFT_CLEAF,        "file" : { "fmt" : RAW, "name" : "output/pft_cleaf.bin"}},
{ "id" : PFT_NLEAF,        "file" : { "fmt" : RAW, "name" : "output/pft_nleaf.bin"}},
{ "id" : PFT_LAIMAX,       "file" : { "fmt" : RAW, "name" : "output/pft_laimax.bin"}},
{ "id" : PFT_CROOT,        "file" : { "fmt" : RAW, "name" : "output/pft_croot.bin"}},
{ "id" : PFT_NROOT,        "file" : { "fmt" : RAW, "name" : "output/pft_nroot.bin"}},
{ "id" : PFT_CSAPW,        "file" : { "fmt" : RAW, "name" : "output/pft_csapw.bin"}},
{ "id" : PFT_NSAPW,        "file" : { "fmt" : RAW, "name" : "output/pft_nsapw.bin"}},
{ "id" : PFT_CHAWO,        "file" : { "fmt" : RAW, "name" : "output/pft_chawo.bin"}},
{ "id" : PFT_NHAWO,        "file" : { "fmt" : RAW, "name" : "output/pft_nhawo.bin"}},
#ifdef WITH_SPITFIRE
{ "id" : MFIREC,           "file" : { "fmt" : RAW, "name" : "output/mfirec.bin"}},
{ "id" : MNFIRE,           "file" : { "fmt" : RAW, "name" : "output/mnfire.bin"}},
{ "id" : MBURNTAREA,       "file" : { "fmt" : RAW, "name" : "output/mburnt_area.bin"}},
#endif
{ "id" : MDISCHARGE,       "file" : { "fmt" : RAW, "name" : "output/mdischarge.bin"}},
{ "id" : MWATERAMOUNT,     "file" : { "fmt" : RAW, "name" : "output/mwateramount.bin"}},
#ifdef DAILY_OUTPUT
{ "id" : D_NPP,            "file" : { "fmt" : RAW, "name" : "output/d_npp.bin"}},
{ "id" : D_GPP,            "file" : { "fmt" : RAW, "name" : "output/d_gpp.bin"}},
{ "id" : D_RH,             "file" : { "fmt" : RAW, "name" : "output/d_rh.bin"}},
{ "id" : D_TRANS,          "file" : { "fmt" : RAW, "name" : "output/d_trans.bin"}},
{ "id" : D_INTERC,         "file" : { "fmt" : RAW, "name" : "output/d_interc.bin"}},
{ "id" : D_EVAP,           "file" : { "fmt" : RAW, "name" : "output/d_evap.bin"}},
#endif
{ "id" : MPET,             "file" : { "fmt" : RAW, "name" : "output/mpet.bin"}},
{ "id" : MALBEDO,          "file" : { "fmt" : RAW, "name" : "output/malbedo.bin"}},
{ "id" : MAXTHAW_DEPTH,    "file" : { "fmt" : RAW, "name" : "output/maxthaw_depth.bin"}},
{ "id" : MSOILTEMP1,       "file" : { "fmt" : RAW, "name" : "output/msoiltemp1.bin"}},
{ "id" : MSOILTEMP2,       "file" : { "fmt" : RAW, "name" : "output/msoiltemp2.bin"}},
{ "id" : MSOILTEMP3,       "file" : { "fmt" : RAW, "name" : "output/msoiltemp3.bin"}},
{ "id" : SOILC_LAYER,      "file" : { "fmt" : RAW, "name" : "output/soilc_layer.bin"}},
{ "id" : AGB,              "file" : { "fmt" : RAW, "name" : "output/agb.bin"}},
{ "id" : MRETURN_FLOW_B,   "file" : { "fmt" : RAW, "name" : "output/mreturn_flow_b.bin"}},
{ "id" : MTRANSP_B,        "file" : { "fmt" : RAW, "name" : "output/mtransp_b.bin"}},
{ "id" : MEVAP_B,          "file" : { "fmt" : RAW, "name" : "output/mevap_b.bin"}},
{ "id" : MINTERC_B,        "file" : { "fmt" : RAW, "name" : "output/mintec_b.bin"}},
{ "id" : ACONV_LOSS_EVAP,  "file" : { "fmt" : RAW, "name" : "output/aconv_loss_evap.bin"}},
{ "id" : ACONV_LOSS_DRAIN, "file" : { "fmt" : RAW, "name" : "output/aconv_loss_drain.bin"}}
/*------------------------ ---------------------- ------------------------------- */
],

"crop_index" : TEMPERATE_CEREALS,  /* CFT for daily output */
"crop_irrigation" : DAILY_RAINFED, /* irrigation flag for daily output */

#else

"output" : [],  /* no output written */

#endif

/*===================================================================*/
/*  V. Run settings section                                          */
/*===================================================================*/

"startgrid" : ALL, /* 27410 67208 60400 all grid cells */

#ifdef CHECKPOINT
"checkpoint_filename" : "restart/restart_checkpoint.lpj", /* filename of checkpoint file */
#endif

#ifndef FROM_RESTART

"nspinup" : 5000,  /* spinup years */
"nspinyear" : 30,  /* cycle length during spinup (yr) */
"firstyear": 1901, /* first year of simulation */
"lastyear" : 1901, /* last year of simulation */
"restart" : false, /* do not start from restart file */
"write_restart" : true, /* create restart file: the last year of simulation=restart-year */
"write_restart_filename" : "restart/restart_1900_nv_stdfire.lpj", /* filename of restart file */
"restart_year": 1900 /* write restart at year */

#else

"nspinup" : 0,  /* spinup years */
"nspinyear" : 30, /* cycle length during spinup (yr)*/
"firstyear": 1901, /* first year of simulation */
"lastyear" : 2011, /* last year of simulation */
"restart" :  true, /* start from restart file */
"restart_filename" : "restart/restart_1900_nv_stdfire.lpj", /* filename of restart file */
"write_restart" : false /* do not create restart file */

#endif
}
